common/nvramparser.cpp

```diff
@@ -296,1548 +316,1225 @@ USTATUS NvramParser::parseNvramVolumeBody(const UModelIndex & index)
-                const EFI_FAULT_TOLERANT_WORKING_BLOCK_HEADER64* ftw64Header = (const EFI_FAULT_TOLERANT_WORKING_BLOCK_HEADER64*)currentPos;
-                if (ftw64Header->WriteQueueSize == 0 || ftw64Header->WriteQueueSize >= 0xFFFFFFFF) {
-                    msg(usprintf("%s: FTW block candidate at offset %Xh skipped, has invalid body size %" PRIX64 "h", __FUNCTION__, localOffset + offset, ftw64Header->WriteQueueSize), index);
-                    continue;
+            
+            // Add free space or padding after all variables, if needed
+            if (entryOffset < storeSize) {
+                UByteArray freeSpace = volumeBody.mid(storeOffset + entryOffset, storeSize - entryOffset);
+                // Add info
+                info = usprintf("Full size: %Xh (%u)", (UINT32)freeSpace.size(), (UINT32)freeSpace.size());
+                
+                // Check that remaining unparsed bytes are actually zeroes
+                if (freeSpace.count('\x00') == freeSpace.size() - 4) { // Free space, 4 last bytes are always CRC32
+                    // Add tree item
+                    model->addItem(entryOffset, Types::FreeSpace, 0, UString("Free space"), UString(), info, UByteArray(), freeSpace, UByteArray(), Fixed, headerIndex);
+                }
+                else {
+                    // Add tree item
+                    model->addItem(entryOffset, Types::Padding, getPaddingType(freeSpace), UString("Padding"), UString(), info, UByteArray(), freeSpace, UByteArray(), Fixed, headerIndex);
                 }
             }
-            else // Unknown header
-                continue;
             
-            // All checks passed, store found
-            break;
+            storeOffset += storeSize - 1;
+            previousStoreEndOffset = storeOffset + 1;
+            continue;
+        } catch (...) {
+            // Parsing failed, try something else
         }
-        else if (readUnaligned(currentPos) == NVRAM_PHOENIX_FLASH_MAP_SIGNATURE_PART1) {// Phoenix SCT flash map
-            UByteArray signature = UByteArray(volume.constData() + offset, NVRAM_PHOENIX_FLASH_MAP_SIGNATURE_LENGTH);
-            if (signature != NVRAM_PHOENIX_FLASH_MAP_SIGNATURE) // Check the whole signature
-                continue;
+not_sysf:
+        // Phoenix Flash Map
+        try {
+            if (volumeBodySize - storeOffset < sizeof(PHOENIX_FLASH_MAP_HEADER)) {
+                // No need to parse further, the rest of the volume is too small
+                goto not_flm;
+            }
+            // Perform initial sanity check
+            const PHOENIX_FLASH_MAP_HEADER* storeHeader = (const PHOENIX_FLASH_MAP_HEADER*)(volumeBody.constData() + storeOffset);
+            if (UByteArray((const char*)storeHeader->Signature, NVRAM_PHOENIX_FLASH_MAP_SIGNATURE_LENGTH) != NVRAM_PHOENIX_FLASH_MAP_SIGNATURE
+                || storeHeader->NumEntries > 113) {
```

Comment:
Could you please at least add the meaning of 113 in a comment?

Reply:
Sure, the map is of 0x1000 size, and 113 is the amount of entries that can still fit in. There's a comment about it in nvram.h, but it needs to be converted into a define.